### PR TITLE
DPE 350 -Start mongod with keyfile and add password

### PR DESCRIPTION
**Pull request by MiaAltieri**
### Prelude 
This is a big PR since authentication must all be done at once. But I want you to note a **lot of the code that was added here is borrowed** from the [k8s-charm](https://github.com/delgod/mongodb-k8s-operator); which code is borrowed is outlined in the "FAQ" section of this. 

This change is also quite large since it is forked off a different PR #36 . Once #36 is merged this should decrease in size

In an attempt to decrease the size of this PR some changes are saved for subsequent PRs, this can be found in the section titled "Subsequent PRs."

### Problem

MongoDB has no security.

### Solution

This solution add security by:
1. start `mongod` with a keyfile 
2. once the replica set is running set up admin user and password

### Context

Initial admin user can be created only through [localhost connection](https://www.mongodb.com/docs/manual/core/localhost-exception/). 

### Release Notes

Changes in this PR.
1. Keyfile added when starting (this cannot be done in install hook since the peer relation databag doesn't exist at this point and information is stored in that databag.)
2. Admin user + password created
3. Functionality to retrieve password
4. Switches from `self._mongo` to `MongoConnection` where relevant. 

### Testing

Unit tests were updated (and some unit tests were removed since their scenarios are tested with integration tests.)

Integration tests were not changed but instead verify that authentication works correctly.

### FAQ
Why is `mongod` sometimes connected to with `MongoConnection` and other times `self._mongo`?
- This is because we are slowly phasing out the file `src/mongod_helpers.py` with `lib/charms/mongodb_libs/v0/mongodb.py`. Eventually usage of `self._mongo` will be removed.

Why are the changes in `src/mongod_helpers.py` so ugly? 
- Because this file is getting deleted soon as we migrate over the library code

What code from this PR is borrowed from the [MongoDB K8s Charm](https://github.com/delgod/mongodb-k8s-operator)?
- everything in `lib/charms/mongodb_libs/v0/mongodb.py`
- `src/charm.py` - `_init_admin_user`
- `src/charm.py` - `app_data`
- `src/charm.py` - `mongodb_config`
- `src/charm.py` - `_initialise_replica_set` (the code within this function was copied from the `on_start` hook from the k8s charm)
- `src/charm.py` - `_on_get_admin_password`
- `src/charm.py` - `_generate_passwords`

### Subsequent PRs
Adding authentication was a huge PR, to prevent this PR from growing any more the following will be added in a subsequent PR:
1. error handling related to creation & permissions of keyfile
2. unit tests for 1.
3. phasing out `self._mongo` and opting for `MongoConnection`
2. removing dead code

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -186,35 +216,86 @@ def _on_install(self, _) -> None:
...
+        pass
 
-        machine_ip = self._unit_ip(self.unit)
-        mongo_config["net"]["bindIp"] = "localhost,{}".format(machine_ip)
-        if "replication" not in mongo_config:
-            mongo_config["replication"] = {}
+    def _instatiate_keyfile(self, event: ops.charm.StartEvent) -> None:
+        # wait for keyFile to be created by leader unit
+        if "keyfile" not in self.app_data:
+            event.defer()
```
Comment (delgod) [inline src/charm.py:270]: should we log an explanation of why we defer?

File: src/charm.py [hunk 2]
```diff
@@ -63,14 +76,30 @@ def __init__(self, *args):
...
+            self.app_data["admin_password"] = generate_password()
+
+        if "keyfile" not in self.app_data:
+            self.app_data["keyfile"] = generate_keyfile()
 
     def _update_hosts(self, event) -> None:
         """Update replica set hosts and remove any unremoved replicas from the config."""
-        if not self.unit.is_leader() or "replset_initialised" not in self._peers.data[self.app]:
+        if "keyfile" not in self.app_data:
+            self._generate_passwords()
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:96]: Perhaps I need a better docstring + function naming, this function is the handler for `leader_elected` events

**Review by marceloneppel — COMMENTED** [review]
LGTM!

File: src/charm.py [hunk 3]
```diff
@@ -203,6 +233,12 @@ def _on_install(self, _) -> None:
                 # part of replicaset
                 "--replSet",
                 "rs0",
+                # TODO enable auth
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:236]: good eye 👀

File: src/charm.py [hunk 4]
```diff
@@ -240,6 +276,9 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
         Args:
             event: The triggering start event.
         """
+        # MongoDB with authentication requires a keyfile
+        self._instatiate_keyfile(event)
+
         # start mongo service
         self.unit.status = MaintenanceStatus("starting MongoDB")
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:283]: I don't see why not! Good idea :)

**Review by delgod — APPROVED** [review]

**Review by paulomach — APPROVED** [review]
LGTM

**Review by marceloneppel — APPROVED** [review]

**Pull request merged**